File: 0128-longest-consecutive-sequence/0128-longest-consecutive-sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        # [100, 4, 200, 1, 3, 2]
        # [ Y.  N   Y.  Y  N. N ]
        s = set(nums)
        longest = 0
        for num in s:
            # is num the start of a sequence
            if num-1 not in s:
                # x = num+1
                length = 1
                n = num
                while n+1 in s:
                    length += 1
                    n += 1
                longest = max(length, longest)
        return longest

File: 0128-longest-consecutive-sequence/0128-longest-consecutive-sequence.java
class Solution {
    public int longestConsecutive(int[] nums) {
        // use set to store elements O(n)
        Set<Integer> s = new HashSet<>();
        int longest = 0;
        for (int num: nums) {
            s.add(num);
        }
        // find start of the sequence and obtain longest sequence O(n)
        for (int num: s) {
            if (!s.contains(num-1)) {
                int length = 1;
                while (s.contains(num+length)) {
                    length++;
                }
                longest = Math.max(longest, length);
            } 
        }
        return longest;
    }
}